x/superfluid/keeper/stake_test.go

```diff
@@ -448,6 +449,84 @@ func (suite *KeeperTestSuite) TestSuperfluidUndelegate() {
+		suite.ctx = suite.ctx.WithBlockTime(startTime)
+		accAddr := suite.app.SuperfluidKeeper.GetLockIdIntermediaryAccountConnection(suite.ctx, lock.ID)
+		intermediaryAcc := suite.app.SuperfluidKeeper.GetIntermediaryAccount(suite.ctx, accAddr)
+		valAddr := intermediaryAcc.ValAddr
+
+		// first we test that SuperfluidUnbondLock would cause error before undelegating
+		err := suite.app.SuperfluidKeeper.SuperfluidUnbondLock(suite.ctx, lock.ID, lock.GetOwner())
+		suite.Require().Error(err)
+
+		// undelegation needs to happen prior to SuperfluidUnbondLock
+		err = suite.app.SuperfluidKeeper.SuperfluidUndelegate(suite.ctx, lock.Owner, lock.ID)
+		suite.Require().NoError(err)
+
+		// check that unbonding synth has been created correctly after undelegation
+		unbondingDuration := suite.app.StakingKeeper.GetParams(suite.ctx).UnbondingTime
+		synthLock, err := suite.app.LockupKeeper.GetSyntheticLockup(suite.ctx, lock.ID, keeper.UnstakingSyntheticDenom(lock.Coins[0].Denom, valAddr))
+		suite.Require().NoError(err)
+		suite.Require().Equal(synthLock.UnderlyingLockId, lock.ID)
+		suite.Require().Equal(synthLock.SynthDenom, keeper.UnstakingSyntheticDenom(lock.Coins[0].Denom, valAddr))
+		suite.Require().Equal(synthLock.EndTime, suite.ctx.BlockTime().Add(unbondingDuration))
+
+		// test SuperfluidUnbondLock
+		unbondLockStartTime := startTime.Add(time.Hour)
+		suite.ctx = suite.ctx.WithBlockTime(unbondLockStartTime)
+		err = suite.app.SuperfluidKeeper.SuperfluidUnbondLock(suite.ctx, lock.ID, lock.GetOwner())
+		suite.Require().NoError(err)
+
+		// check that SuperfluidUnbondLock makes underlying lock start unlocking
+		// we run WithdrawAllMaturedLocks to ensure that lock isn't getting finished immediately
+		suite.app.LockupKeeper.WithdrawAllMaturedLocks(suite.ctx)
+		updatedLock, err := suite.app.LockupKeeper.GetLockByID(suite.ctx, lock.ID)
+		suite.Require().NoError(err)
+		suite.Require().False(lock.IsUnlocking())
+		suite.Require().True(updatedLock.IsUnlocking())
+
+		// test that synth lock finish does not mean underlying lock is finished
+		suite.ctx = suite.ctx.WithBlockTime((startTime.Add(unbondingDuration)))
+		suite.app.LockupKeeper.DeleteAllMaturedSyntheticLocks(suite.ctx)
+		suite.app.LockupKeeper.WithdrawAllMaturedLocks(suite.ctx)
+		_, err = suite.app.LockupKeeper.GetSyntheticLockup(suite.ctx, lock.ID, keeper.UnstakingSyntheticDenom(lock.Coins[0].Denom, valAddr))
+		suite.Require().Error(err)
+		updatedLock, err = suite.app.LockupKeeper.GetLockByID(suite.ctx, lock.ID)
+		suite.Require().NoError(err)
+		suite.Require().True(updatedLock.IsUnlocking())
+
+		// test after SuperfluidUnbondLock + lockup unbonding duration, lock is finished and does not exist
+		suite.ctx = suite.ctx.WithBlockTime(unbondLockStartTime.Add(unbondingDuration))
+		suite.app.LockupKeeper.WithdrawAllMaturedLocks(suite.ctx)
+		updatedLock, err = suite.app.LockupKeeper.GetLockByID(suite.ctx, lock.ID)
+		fmt.Println(updatedLock)
+		suite.Require().Error(err)
+
```

Comment:
We need to check that the users underlying balance is increased after the withdraw. (exactly as expected)

There was previously weird bugs where the locks underlying denom was changed on the synthetic lockup removal.